python/glow/wgr/linear_model/logistic_model.py

```diff
@@ -0,0 +1,238 @@
+            covdf: pd.DataFrame) -> (DataFrame, DataFrame):
+        """
+        Fits a logistic regression model, represented by a Spark DataFrame containing coefficients for each of the ridge
+        alpha parameters, for each block in the starting matrix, for each label in the target labels, as well as a
+        Spark DataFrame containing the optimal ridge alpha value for each label.
+
+        Args:
+            blockdf : Spark DataFrame representing the beginning block matrix X
+            labeldf : Pandas DataFrame containing the target labels used in fitting the ridge models
+            sample_blocks : Dict containing a mapping of sample_block ID to a list of corresponding sample IDs
+            covdf : Pandas DataFrame containing covariates to be included in every model in the stacking
+                ensemble (optional).
+
+        Returns:
+            Two Spark DataFrames, one containing the model resulting from the fitting routine and one containing the
+            results of the cross validation procedure.
+        """
+        validate_inputs(labeldf, covdf, 'binary')
+        map_key_pattern = ['sample_block', 'label', 'alpha_name']
+        reduce_key_pattern = ['header_block', 'header', 'label', 'alpha_name']
+        model_key_pattern = ['sample_block', 'label', 'alpha_name']
+        score_key_pattern = ['sample_block', 'label']
+        p0_dict = {k: v for k, v in zip(labeldf.columns, labeldf.sum(axis=0) / labeldf.shape[0])}
+        metric = 'log_loss'
+
+        if not self.alphas:
+            self.alphas = generate_alphas(blockdf)
+
+        map_udf = pandas_udf(
+            lambda key, pdf: map_irls_eqn(key, map_key_pattern, pdf, labeldf, sample_blocks, covdf,
+                                          p0_dict, self.alphas), irls_eqn_struct,
+            PandasUDFType.GROUPED_MAP)
+
+        reduce_udf = pandas_udf(lambda key, pdf: reduce_irls_eqn(key, reduce_key_pattern, pdf),
+                                irls_eqn_struct, PandasUDFType.GROUPED_MAP)
+
+        model_udf = pandas_udf(
+            lambda key, pdf: solve_irls_eqn(key, model_key_pattern, pdf, labeldf, self.alphas),
+            model_struct, PandasUDFType.GROUPED_MAP)
+
+        score_udf = pandas_udf(
+            lambda key, pdf: score_models(key, score_key_pattern, pdf, labeldf, sample_blocks, self.
+                                          alphas, covdf, metric), cv_struct,
+            PandasUDFType.GROUPED_MAP)
+
+        modeldf = blockdf.drop('alpha') \
+            .withColumn('alpha_name', f.explode(f.array([f.lit(n) for n in self.alphas.keys()]))) \
+            .groupBy(map_key_pattern) \
+            .apply(map_udf) \
+            .groupBy(reduce_key_pattern) \
+            .apply(reduce_udf) \
+            .groupBy(model_key_pattern) \
+            .apply(model_udf) \
+            .withColumn('alpha_label_coef', f.expr('struct(alphas[0] AS alpha, labels[0] AS label, coefficients[0] AS coefficient)')) \
```

Comment:
This goes with my comment in `logistic_udfs.py`, but if these are guaranteed to be singletons, they shouldn't be wrapped in arrays.

Reply:
See discussion below